doc/sphinx/proof-engine/ltac2.rst

```diff
@@ -476,14 +504,138 @@ In Ltac2, we have the following backtracking primitives, defined in the
+          let n := to_stream v () in
+          try (Message.print (Message.of_constr n); fail).
+          (** Prints 0, 1, 2. *)
+
+     .. coqtop:: all abort
+
+This illustrates that one can view thunks as lazy lists of results,
+where `zero` is the empty list and `plus` is list concatenation, while
+`case` is pattern-matching.
+
+It is also possible to create an infinite stream of values using a thunk.
+
+  .. example::
+
+     .. coqtop:: in
+
+        Ltac2 int_seq () :=
+          let rec succ thunk_n :=
+          Control.plus thunk_n (fun _ => succ (fun _ => Int.add (thunk_n ()) 1))
+                in succ (fun _ => 0).
+
+            Goal nat.
+                let n := int_seq () in
+                if Int.lt n 10 then (Message.print (Message.of_int n); fail)
+                else ().
+                (** Prints 0, 1, 2... 9 . *)
+
+     .. coqtop:: all abort
+
+The relationship between `to_stream l` and the "downstream"
+computation is a naive one. The downstream computation either
+succeeds, in which case control never returns to `to_stream l`; or it
+fails, in which case `to_stream l` naively returns the next element in
+the stream. The same criticism applies to `int_seq ()`. We would like
+to engineer more sophisticated error handling; in a situation `t_1;
+t_2`, if `t_2` fails, it is desirable for `t_2` to be able to
+communicate the manner of failure back upstream to `t_1`, so that
+`t_1` can intelligently respond with a new value (and a new change in
+the proof state) in an appropriate way.
+
+The most general way of doing this is to encode a state machine into `t_1`, so that a
+downstream tactic `t_2` which fails can communicate useful information back
+upstream through the exception it raises, and `t_1` can choose which value
+to return next based on the exception and its current state.
+
+  .. example::
+
+     .. coqtop:: in
+
+        Ltac2 state_transition_thunk (init_state : 's) (f : 's * exn -> 's * 'a)
+                (init_guess: 'a) :=
+            let rec succ state e :=
+                let (next_state, next_guess) :=  f (state, e) in
+                Control.plus (fun _ => next_guess)
+                    (fun next_exn => succ next_state next_exn)
+            in Control.plus (fun _ => init_guess)
+                   (fun init_exn => succ init_state init_exn).
+
+
+The function `case` allows the user to implement similar backtracking
+control flow functionality to what is baked into Ltac2 with the
+semicolon operator and let bindings.
+
+To illustrate, we will show how to implement the semicolon operator
+ourselves directly using `case`:
+
+.. example::
+
+   .. coqtop:: in
+
+      Ltac2 rec concat (tac1 : unit -> unit) (tac2 : unit -> 'a) : 'a :=
+        match Control.case tac1 with
+        | Err e => Control.zero e
+        | Val ( _ , f ) =>
+             Control.plus tac2 (fun e' =>
+             concat (fun () => f e') (fun _ => tac2 ()))
+        end.
```

Comment:
Ltac2 seems very general. I did not expect `tac1` to be evaluated knowing how `tac2` failed, i.e. `f e'` or then we expect in the traditional model. In other words, I was expecting `f` to be constant.